tools/tester.rb

```diff
@@ -248,6 +298,16 @@ def run_tests
 
   ret = failed == 0
 
+
+  if faillist.any?
+    puts "failed tests:"
+
+    faillist.each { |test_case| test_case.results
```

Comment:
This is pretty hard to understand for me. AFAIR Ruby convention is to use begin/end instead of {} when the block spans multiple lines. I would suggest to write this something like this:

```ruby
for test_case in faillist
  for run_name, run_result in test_case.results
    next if run_result == true
    puts "  #{test_case.file}.#{run_name}"
  end
end
```